SYCL/InvokeSimd/fig_4_13_nd_range_matrix_multiply_accessor_get_pointer.cpp

```diff
@@ -0,0 +1,241 @@
+
+  auto dev = q.get_device();
+  std::cout << "Running on " << dev.get_info<info::device::name>() << "\n";
+  auto ctxt = q.get_context();
+
+  // Initialize input and output memory on the host
+  // NOTE: In this program, the value of B is not really important (i.e., it is
+  // more or less an arbitrary value); the only reason we are using explicit ND
+  // range kernels (i.e., nd_range, nd_item) in this program is because we need
+  // to extract the subgroup from the nd_item in order to pass it to
+  // invoke_simd(), which requires a subgroup argument. A local (execution)
+  // range is defined and consequently workitems in the global execution range
+  // are grouped into workgroups, however, we do not use any this grouping or
+  // any workgroup-specific functionality for this computation (i.e., workgroup
+  // local memory, work-item communication or synchronization within a
+  // workgroup, etc.).
+  // constexpr size_t B = 64; // 256 / 64 = 4. The matrix is divided into 4 x 4
+  // = 16 workgroups of size 64 x 64 = 4,096 workitems each.
+  // Visualize each workgroup as a B x B sub-matrix.
+  constexpr size_t B = 4; // The matrix is divided into 4 x 4 = 16 workgroups.
+  constexpr unsigned GroupSize =
+      B * B; // B * B = B^2 workitems in each workgroup.
+  // constexpr unsigned GroupSize = 4 * VL; // VL is the subgroup size, 16; each
+  // workgroup will contain 4 subgroups of 16 workitems each.
+
+  // Create and initialize our N x N matrices. Remember that multidimensionality
+  // is a programmer convenience implemented on top of an underlying
+  // one-dimensional space (pg. 94).
+  std::vector<int> a(N * N), b(N * N), c(N * N);
+  std::fill(a.begin(), a.end(), 1);
+  std::fill(b.begin(), b.end(), 2);
+  std::fill(c.begin(), c.end(), 0);
+
+  {
+    // Create buffers associated with inputs and output
+    buffer<int, 2> a_buf(a.data(), range<2>(N, N)),
+        b_buf(b.data(), range<2>(N, N)), c_buf(c.data(), range<2>(N, N));
+
+    try {
+      // Submit the kernel to the queue
+      auto e = q.submit([&](handler &h) {
+        accessor acc_a{a_buf, h};
+        accessor acc_b{b_buf, h};
+        accessor acc_c{c_buf, h};
+
+        // START CODE SNIP
+        range global{N, N};
+        range local{B, B};
+        h.parallel_for<class Test>(
+            nd_range{global, local},
+            [=](nd_item<2> ndi) [[intel::reqd_sub_group_size(VL)]] {
+              int row_index = ndi.get_global_id(0);
+              int column_index = ndi.get_global_id(1);
+
+              if constexpr (use_invoke_simd) {
+                // Currently, uniform{accessor::get_pointer()} returns incorrect
+                // results if the pointer returned is not explicitly cast to an
+                // appropriate data type (i.e., int *, etc.). "To me this looks
+                // suspicious, maybe a bug. Might be related to address space
+                // discrepancy." - Konst.
```

Comment:
I suggest to at least mark failing test cases with TODO. The best would be to extract them into separate XFAIL tests.
```suggestion
                // TODO FIXME: Currently, uniform{accessor::get_pointer()} returns incorrect
                // results if the pointer returned is not explicitly cast to an
                // appropriate data type (i.e., int *, etc.).
```

Reply:
It looks like this comment is outdated. I looked at earlier revisions of this test and static_cast was removed at some point, but comment remained.